### ISSUE TITLE
[Submission] 44 params - Tom Bukic

### ISSUE THREAD

**Issue by tbukic** [anadim/AdderBoard#78]
### Author

tbukic

### Unique Parameter Count

44

### Accuracy

100.00% on 10,010 tests (verify.py seed=2025: 10K random + 10 edge cases)

### Method

Trained (SGD/Adam)

### Architecture

1L decoder-only transformer + circular arc embedding, d=3, 1h/1kv, hd=4, ff=2, RoPE theta=3, SwiGLU, RMSNorm

### Key Tricks

- Circular arc embedding (3 params instead of 30): digit tokens mapped via A*cos(start + stride*i)
- K = Q (key shares query weights exactly, 0 extra params — replaces 12-param key projection)
- V = Q (value shares query weights exactly, 0 extra params — replaces 12-param value projection)
- Tied O=Q^T (output projection = transpose of query projection, saves 12 params)Also
- All 3 RMSNorms shared (ln1, ln2, final_norm share weights, saves 6 params)
- Tied lm_head to dynamic embedding table
- RoPE positional encoding (zero trainable params, theta=3)
- QK norms (RMSNorm on Q and K, stabilizes training)
- SwiGLU MLP at ff=2
- Curriculum learning + Grokfast-EMA gradient filter (alpha=0.98, lambda=1.0) + adaptive weight decay
- Pure grokking — no targeted fine-tuning needed

### Link to Code

https://github.com/tbukic/M10S-Transformer/blob/main/submissions/submission_44p.py

### Verification Output

```shell
Model: M10S-44p
Author: Tom Bukic
Parameters (unique): 44
Architecture: 1L decoder-only transformer + circular arc embed, d=3, 1h/1kv, hd=4, ff=2, K=Q, V=Q, tieQO, shnorm

Results: 10010/10010 correct (100.00%)
Status: QUALIFIED (threshold: 99%)
```

### Additional Notes

- Parameter budget (44 total):
  - 3 circular arc embedding (A, start, stride)
  - 3 ln1 RMSNorm (shared with ln2 and final_norm via shnorm)
  - 12 q_proj (shared with k_proj as K=Q, v_proj as V=Q, and o_proj via tieQO)
  - 4+4 QK norms (q_norm + k_norm)
  - 6 up_proj
  - 6 down_proj
  - 6 gate_proj
- Training pipeline: 200K cosine LR sweep (seed=27, 83.4%) → 64K adamw+gf continuation → natural grokking at step 264K
- No targeted fine-tuning needed — the model reached 100% on 10K test set purely through grokking
- Extreme weight sharing: Q, K, V, and O all share the same 12-param projection (K=Q, V=Q, O=Q^T)
- Grokfast-EMA was essential: amplifies slow-varying gradient components for generalization
